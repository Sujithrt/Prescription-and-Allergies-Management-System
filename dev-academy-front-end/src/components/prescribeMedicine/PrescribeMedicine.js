import React, { useState } from 'react';
import Select, { components } from 'react-select';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import IconButton from '@material-ui/core/IconButton';
import AddIcon from '@material-ui/icons/Add';
import { retrieveMedicines, prescribeMedicine } from '../../api/PnamsService';
import Button from 'react-bootstrap/Button';
import TextField from '@mui/material/TextField';
import { Alert } from 'react-bootstrap';
import './../../pages/patientProfile/patientProfile.css';
import { useLocation, useHistory } from 'react-router-dom';
import ReactTooltip from 'react-tooltip';
import WarningAmberIcon from '@mui/icons-material/WarningAmber';
import _ from 'lodash';

/**
 * Component that allows user to prescribe medicine to a patient
 * 
 * @returns 
 */
export default function PrescribeMedicine(props) {

    const search = useLocation().search;
    const history = useHistory();
    const { Option } = components;
    const [inputMedicineFields, setInputMedicineFields] = useState([
        { medicine: "", dose: "" }
    ]);
    const [error, setError] = useState("");
    const [medicineOptions, setMedicineOptions] = useState([]);

    /**
     * If encounterId is present in the query param, function will extract it and return the encounterId
     * @returns
     */
    const getEncounterIdFromQueryParam = () => {
        const encounterId = new URLSearchParams(search).get('encounterId');
        return encounterId
    }

    /**
     * Handles change in values of Select Medicine Search Bar
     * @param {Prescribe Medicine Form Index} index 
     * @param {onChange} event 
     */
    const handleChangeInput = (index, event) => {
        const values = [...inputMedicineFields];
        if (null !== event) {
            values[index]['medicine'] = event;

        } else {
            values[index]['medicine'] = "";
        }
        setInputMedicineFields(values);
    }

    /**
     * Handles change in values of dose input field
     * @param {Prescribe Medicine Form Index} index 
     * @param {onChange} event 
     */
    const handleChangeDose = (index, event) => {
        const values = [...inputMedicineFields];
        values[index]['dose'] = event.target.value;
        setInputMedicineFields(values);
    }

    /**
     * Adds a new Add Medicine Form which allows user to prescribe more medicines
     * @param {onClick} index 
     */
    const handleAddMedicine = (e) => {
        setInputMedicineFields([...inputMedicineFields, { medicine: "", dose: "" }]);
    }

    /**
     * Allows user to remove fields of Prescribe Medicine Form
     * @param {onClick} index 
     */
    const handleRemoveMedicine = (index) => {
        setError("");
        const values = [...inputMedicineFields];
        values.splice(index, 1);
        setInputMedicineFields(values);
    }

    /**
     * Allows user to prescribe medicine for a patient
     * @param {onClick} event 
     */
    const handlePrescribe = (event) => {
        event.preventDefault();
        const patientId = props.patient.id;
        let flag = 0;
        for (let i = 0; i < inputMedicineFields.length; i++) {
            if (inputMedicineFields[i].medicine === "" || inputMedicineFields[i].medicine.label === "" ||
                inputMedicineFields[i].medicine.value <= 0 || inputMedicineFields[i].dose === "" ||
                inputMedicineFields[i].medicine === null || inputMedicineFields[i].medicine.label === null
                || inputMedicineFields[i].medicine.value === null || inputMedicineFields[i].dose === null) {
                flag = 1;
                break;
            }
        }
        if (flag === 1) {
            setError("Please fill all fields");
        } else {
            setError("");
            if (getEncounterIdFromQueryParam() === null) {
                const medicinesData = {
                    encounterId: 0,
                    prescribedMedicineList: inputMedicineFields
                };
                prescribeMedicine(patientId, medicinesData)
                    .then((response) => {
                        window.location.search = "encounterId=" + response.data[0].encounter.id;
                        setInputMedicineFields([{ medicine: "", dose: "" }]);
                        setError("");
                    })
                    .catch((error) => {
                        if (error.response.status === 400) {
                            setError("Please check your input");
                        } else if (error.response.status === 409) {
                            setError("Medicine already prescribed in current encounter");
                        } else {
                            history.push("/error");
                        }
                    });
            } else {
                const medicinesData = {
                    encounterId: getEncounterIdFromQueryParam(),
                    prescribedMedicineList: inputMedicineFields
                };
                prescribeMedicine(patientId, medicinesData)
                    .then((response) => {
                        window.location.reload();
                        setInputMedicineFields([{ medicine: "", dose: "" }]);
                        setError("");
                    })
                    .catch((error) => {
                        if (error.response.status === 400) {
                            setError("Please check your input");
                        } else if (error.response.status === 409) {
                            setError("Medicine already prescribed in current encounter");
                        } else {
                            history.push("/error");
                        }
                    });
            }
        }
    }

    /**
     * Allows user to search for a medicine and retrieves medicines from database based on search text
     * @param searchText 
     */
    const searchMedicine = (searchText) => {
        _.debounce(loadMedicines, 1000)(searchText);
    }

    /**
     * Loads list of medicines from the database and shows it to the user
     * @param {onChange} searchText 
     * @returns 
     */
    const loadMedicines = async (searchText) => {
        retrieveMedicines(searchText, props.patient.id)
            .then((response) =>
                setMedicineOptions(response.data.map(i => ({ label: i.brandName, value: i.id, cause: i.cause })))
            )
            .catch((error) => {
                if (error.response.status === 500) {
                    history.push("/error");
                }
            })
    }

    /**
     * Customize select search component to shows warning labels for loaded medicines
     * @param props 
     * @returns 
     */
    const customSelectOption = (props) => (
        <Option {...props}>
            <span>{props.data.label}</span>
            <span className="warning" data-tip={props.data.cause}>
                {(props.data.cause) ?
                    <WarningAmberIcon fontSize="small" /> :
                    <></>}
            </span>
            <ReactTooltip place="top" effect="solid" />
        </Option>
    )

    return (
        <>
            <span className="userUpdateTitle">Medicines</span>
            <IconButton
                data-testid="addMedicineField"
                onClick={handleAddMedicine}>
                <AddIcon />
            </IconButton>
            <br />
            <form onSubmit={handlePrescribe}>
                <Container>
                    {!(error === "") && <Alert variant="danger">{error}</Alert>}
                    {
                        inputMedicineFields.map((inputMedicineField, index) => (
                            <Row key={index}>
                                <Select
                                    name="medicines"
                                    data-testid="medicine"
                                    options={medicineOptions}
                                    onChange={(event) => handleChangeInput(index, event)}
                                    value={inputMedicineFields[index].medicine}
                                    onInputChange={searchMedicine}
                                    className="search-bar"
                                    classNamePrefix="select"
                                    components={{ Option: customSelectOption }}
                                    placeholder="Select medicine..."
                                    isClearable={true}
                                />
                                <TextField
                                    data-testid="doseField"
                                    label="Dose"
                                    variant="standard"
                                    className="doseInput"
                                    name="dose"
                                    value={inputMedicineFields[index].dose}
                                    onChange={(event) => handleChangeDose(index, event)}
                                />
                                <div style={{ marginTop: "10px", marginLeft: "10px" }}>
                                    <Button data-testid="deleteMedicineField" variant="danger" onClick={() => handleRemoveMedicine(index)}>Delete</Button>
                                </div>
                            </Row>
                        ))
                    }
                    {!(inputMedicineFields && inputMedicineFields.length) ? <></> : <Button data-testid="prescribeButton" style={{ marginTop: "10px" }} variant="primary" onClick={handlePrescribe}>Prescribe</Button>}
                </Container>
            </form>
        </>
    )
}
