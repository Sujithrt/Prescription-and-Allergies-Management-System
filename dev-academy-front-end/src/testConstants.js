export const patient = {
    "id": 1001,
    "firstName": "Leanne",
    "lastName": "Graham",
    "gender": "Male",
    "age": 21,
    "phoneNumber": 9123456780,
    "email": "leanne.graham@gmail.com",
    "city": "Bangalore",
    "state": "Karnataka",
    "medicalConditions": [
        {
            "id": 1,
            "name": "Cephalalgia (headache)"
        },
        {
            "id": 3,
            "name": "Otitis externa (swimmer's ear)"
        },
        {
            "id": 9,
            "name": "Contusion (bruise)"
        }
    ],
    "allergies": [
        {
            "id": 7,
            "name": "Aspirin Allergy",
            "reaction": "Tiredness",
            "severity": "MODERATE",
            "status": "ACTIVE",
            "ingredient": {
                "id": 4,
                "name": "Aspirin"
            },
            "encounter": {
                "id": 10,
                "date": "2021-10-26",
                "patient": {
                    "id": 1001,
                    "firstName": "Leanne",
                    "lastName": "Graham",
                    "gender": "Male",
                    "age": 21,
                    "phoneNumber": 9123456780,
                    "email": "leanne.graham@gmail.com",
                    "city": "Bangalore",
                    "state": "Karnataka",
                    "medicalConditions": [
                        {
                            "id": 1,
                            "name": "Cephalalgia (headache)"
                        },
                        {
                            "id": 3,
                            "name": "Otitis externa (swimmer's ear)"
                        },
                        {
                            "id": 9,
                            "name": "Contusion (bruise)"
                        }
                    ]
                },
                "user": {
                    "id": 1,
                    "username": "John",
                    "password": "$2a$10$uS1MVfXzQkqbY7XDruby/OySJp3yjLmo4QGDSiE30GoN0wGb7DCqS",
                    "email": "john.cage@gmail.com",
                    "enabled": true,
                    "role": "Physician",
                    "authorities": [
                        {
                            "id": 1,
                            "roleCode": "USER",
                            "roleDescription": "User role",
                            "authority": "USER"
                        },
                        {
                            "id": 2,
                            "roleCode": "ADMIN",
                            "roleDescription": "Admin role",
                            "authority": "ADMIN"
                        }
                    ],
                    "credentialsNonExpired": true,
                    "accountNonExpired": true,
                    "accountNonLocked": true
                }
            },
            "patient": {
                "id": 1001,
                "firstName": "Leanne",
                "lastName": "Graham",
                "gender": "Male",
                "age": 21,
                "phoneNumber": 9123456780,
                "email": "leanne.graham@gmail.com",
                "city": "Bangalore",
                "state": "Karnataka",
                "medicalConditions": [
                    {
                        "id": 1,
                        "name": "Cephalalgia (headache)"
                    },
                    {
                        "id": 3,
                        "name": "Otitis externa (swimmer's ear)"
                    },
                    {
                        "id": 9,
                        "name": "Contusion (bruise)"
                    }
                ]
            }
        },
        {
            "id": 8,
            "name": "Paracetamol",
            "reaction": "sdfd",
            "severity": "MILD",
            "status": "ACTIVE",
            "ingredient": {
                "id": 1,
                "name": "Paracetamol"
            },
            "encounter": {
                "id": 13,
                "date": "2021-10-26",
                "patient": {
                    "id": 1001,
                    "firstName": "Leanne",
                    "lastName": "Graham",
                    "gender": "Male",
                    "age": 21,
                    "phoneNumber": 9123456780,
                    "email": "leanne.graham@gmail.com",
                    "city": "Bangalore",
                    "state": "Karnataka",
                    "medicalConditions": [
                        {
                            "id": 1,
                            "name": "Cephalalgia (headache)"
                        },
                        {
                            "id": 3,
                            "name": "Otitis externa (swimmer's ear)"
                        },
                        {
                            "id": 9,
                            "name": "Contusion (bruise)"
                        }
                    ]
                },
                "user": {
                    "id": 1,
                    "username": "John",
                    "password": "$2a$10$uS1MVfXzQkqbY7XDruby/OySJp3yjLmo4QGDSiE30GoN0wGb7DCqS",
                    "email": "john.cage@gmail.com",
                    "enabled": true,
                    "role": "Physician",
                    "authorities": [
                        {
                            "id": 1,
                            "roleCode": "USER",
                            "roleDescription": "User role",
                            "authority": "USER"
                        },
                        {
                            "id": 2,
                            "roleCode": "ADMIN",
                            "roleDescription": "Admin role",
                            "authority": "ADMIN"
                        }
                    ],
                    "credentialsNonExpired": true,
                    "accountNonExpired": true,
                    "accountNonLocked": true
                }
            },
            "patient": {
                "id": 1001,
                "firstName": "Leanne",
                "lastName": "Graham",
                "gender": "Male",
                "age": 21,
                "phoneNumber": 9123456780,
                "email": "leanne.graham@gmail.com",
                "city": "Bangalore",
                "state": "Karnataka",
                "medicalConditions": [
                    {
                        "id": 1,
                        "name": "Cephalalgia (headache)"
                    },
                    {
                        "id": 3,
                        "name": "Otitis externa (swimmer's ear)"
                    },
                    {
                        "id": 9,
                        "name": "Contusion (bruise)"
                    }
                ]
            }
        },
        {
            "id": 13,
            "name": "Ibuprofin Allergy",
            "reaction": "Increased Perspiration",
            "severity": "MODERATE",
            "status": "ACTIVE",
            "ingredient": {
                "id": 3,
                "name": "Ibuprofen"
            },
            "encounter": {
                "id": 20,
                "date": "2021-10-29",
                "patient": {
                    "id": 1001,
                    "firstName": "Leanne",
                    "lastName": "Graham",
                    "gender": "Male",
                    "age": 21,
                    "phoneNumber": 9123456780,
                    "email": "leanne.graham@gmail.com",
                    "city": "Bangalore",
                    "state": "Karnataka",
                    "medicalConditions": [
                        {
                            "id": 1,
                            "name": "Cephalalgia (headache)"
                        },
                        {
                            "id": 3,
                            "name": "Otitis externa (swimmer's ear)"
                        },
                        {
                            "id": 9,
                            "name": "Contusion (bruise)"
                        }
                    ]
                },
                "user": {
                    "id": 1,
                    "username": "John",
                    "password": "$2a$10$uS1MVfXzQkqbY7XDruby/OySJp3yjLmo4QGDSiE30GoN0wGb7DCqS",
                    "email": "john.cage@gmail.com",
                    "enabled": true,
                    "role": "Physician",
                    "authorities": [
                        {
                            "id": 1,
                            "roleCode": "USER",
                            "roleDescription": "User role",
                            "authority": "USER"
                        },
                        {
                            "id": 2,
                            "roleCode": "ADMIN",
                            "roleDescription": "Admin role",
                            "authority": "ADMIN"
                        }
                    ],
                    "credentialsNonExpired": true,
                    "accountNonExpired": true,
                    "accountNonLocked": true
                }
            },
            "patient": {
                "id": 1001,
                "firstName": "Leanne",
                "lastName": "Graham",
                "gender": "Male",
                "age": 21,
                "phoneNumber": 9123456780,
                "email": "leanne.graham@gmail.com",
                "city": "Bangalore",
                "state": "Karnataka",
                "medicalConditions": [
                    {
                        "id": 1,
                        "name": "Cephalalgia (headache)"
                    },
                    {
                        "id": 3,
                        "name": "Otitis externa (swimmer's ear)"
                    },
                    {
                        "id": 9,
                        "name": "Contusion (bruise)"
                    }
                ]
            }
        },
        {
            "id": 14,
            "name": "Paracetamol Allergy",
            "reaction": "Increased Perspiration",
            "severity": "MODERATE",
            "status": "ACTIVE",
            "ingredient": {
                "id": 3,
                "name": "Ibuprofen"
            },
            "encounter": {
                "id": 1,
                "date": "2021-10-26",
                "patient": {
                    "id": 1003,
                    "firstName": "Clementine",
                    "lastName": "Bauch",
                    "gender": "Female",
                    "age": 56,
                    "phoneNumber": 9123456782,
                    "email": "clementine.bauch@gmail.com",
                    "city": "Mysore",
                    "state": "Karnataka",
                    "medicalConditions": [
                        {
                            "id": 23,
                            "name": "Gestation (pregnancy) "
                        }
                    ]
                },
                "user": {
                    "id": 1,
                    "username": "John",
                    "password": "$2a$10$uS1MVfXzQkqbY7XDruby/OySJp3yjLmo4QGDSiE30GoN0wGb7DCqS",
                    "email": "john.cage@gmail.com",
                    "enabled": true,
                    "role": "Physician",
                    "authorities": [
                        {
                            "id": 1,
                            "roleCode": "USER",
                            "roleDescription": "User role",
                            "authority": "USER"
                        },
                        {
                            "id": 2,
                            "roleCode": "ADMIN",
                            "roleDescription": "Admin role",
                            "authority": "ADMIN"
                        }
                    ],
                    "credentialsNonExpired": true,
                    "accountNonExpired": true,
                    "accountNonLocked": true
                }
            },
            "patient": {
                "id": 1001,
                "firstName": "Leanne",
                "lastName": "Graham",
                "gender": "Male",
                "age": 21,
                "phoneNumber": 9123456780,
                "email": "leanne.graham@gmail.com",
                "city": "Bangalore",
                "state": "Karnataka",
                "medicalConditions": [
                    {
                        "id": 1,
                        "name": "Cephalalgia (headache)"
                    },
                    {
                        "id": 3,
                        "name": "Otitis externa (swimmer's ear)"
                    },
                    {
                        "id": 9,
                        "name": "Contusion (bruise)"
                    }
                ]
            }
        }
    ]
};

export const user = {
    "id": 1,
    "username": "John",
    "email": "john.cage@gmail.com",
    "role": "Physician",
    "firstName": "John",
    "lastName": "Cage",
    "speciality": "Cardiologist",
    "city": "Bangalore",
    "state": "Karnataka",
    "dob": "1975-06-12T18:30:00.000+00:00",
    "authorities": [
        {
            "id": 1,
            "roleCode": "USER",
            "roleDescription": "User role",
            "authority": "USER"
        },
        {
            "id": 2,
            "roleCode": "ADMIN",
            "roleDescription": "Admin role",
            "authority": "ADMIN"
        }
    ]
};

export const prescription = [
    {
        "id": {
            "encounterId": 1,
            "medicineId": 1
        },
        "encounter": {
            "id": 29,
            "date": "2021-11-07",
            "patient": {
                "id": 1009,
                "firstName": "Glenna",
                "lastName": "Reichert",
                "gender": "Female",
                "age": 65,
                "phoneNumber": 9123456788,
                "email": "glenna.reichert@gmail.com",
                "city": "Kolkata",
                "state": "West Bengal",
                "medicalConditions": [
                    {
                        "id": 6,
                        "name": "Tussis (cough)"
                    },
                    {
                        "id": 7,
                        "name": "Pyrexia (fever)"
                    }
                ]
            },
            "user": {
                "id": 1,
                "username": "John",
                "password": "$2a$10$uS1MVfXzQkqbY7XDruby/OySJp3yjLmo4QGDSiE30GoN0wGb7DCqS",
                "email": "john.cage@gmail.com",
                "enabled": true,
                "role": "Physician",
                "authorities": [
                    {
                        "id": 1,
                        "roleCode": "USER",
                        "roleDescription": "User role",
                        "authority": "USER"
                    },
                    {
                        "id": 2,
                        "roleCode": "ADMIN",
                        "roleDescription": "Admin role",
                        "authority": "ADMIN"
                    }
                ],
                "credentialsNonExpired": true,
                "accountNonExpired": true,
                "accountNonLocked": true
            }
        },
        "medicine": {
            "id": 1,
            "genericName": "Paracetamol",
            "brandName": "Dolo",
            "category": {
                "id": 1,
                "name": "Antipyretics"
            },
            "ingredients": [
                {
                    "id": 1,
                    "name": "Paracetamol"
                }
            ]
        },
        "dosage": "1 time a day"
    }
];

export const allergy = {
    "id": 19,
    "name": "Paracetamol Allergy",
    "reaction": "Tiredness",
    "severity": "MILD",
    "status": "ACTIVE",
    "ingredient": {
        "id": 1,
        "name": "Paracetamol"
    },
    "encounter": {
        "id": 28,
        "date": "2021-11-07",
        "patient": {
            "id": 1009,
            "firstName": "Glenna",
            "lastName": "Reichert",
            "gender": "Female",
            "age": 65,
            "phoneNumber": 9123456788,
            "email": "glenna.reichert@gmail.com",
            "city": "Kolkata",
            "state": "West Bengal",
            "medicalConditions": [
                {
                    "id": 6,
                    "name": "Tussis (cough)"
                },
                {
                    "id": 7,
                    "name": "Pyrexia (fever)"
                }
            ]
        },
        "user": {
            "id": 1,
            "username": "John",
            "password": "$2a$10$uS1MVfXzQkqbY7XDruby/OySJp3yjLmo4QGDSiE30GoN0wGb7DCqS",
            "email": "john.cage@gmail.com",
            "enabled": true,
            "role": "Physician",
            "authorities": [
                {
                    "id": 1,
                    "roleCode": "USER",
                    "roleDescription": "User role",
                    "authority": "USER"
                },
                {
                    "id": 2,
                    "roleCode": "ADMIN",
                    "roleDescription": "Admin role",
                    "authority": "ADMIN"
                }
            ],
            "credentialsNonExpired": true,
            "accountNonExpired": true,
            "accountNonLocked": true
        }
    },
    "patient": {
        "id": 1009,
        "firstName": "Glenna",
        "lastName": "Reichert",
        "gender": "Female",
        "age": 65,
        "phoneNumber": 9123456788,
        "email": "glenna.reichert@gmail.com",
        "city": "Kolkata",
        "state": "West Bengal",
        "medicalConditions": [
            {
                "id": 6,
                "name": "Tussis (cough)"
            },
            {
                "id": 7,
                "name": "Pyrexia (fever)"
            }
        ]
    }
};

export const recentPatients = [
    {
        "id": 1009,
        "firstName": "Glenna",
        "lastName": "Reichert",
        "age": 65,
        "gender": "Female",
        "date": "2021-11-09"
    },
    {
        "id": 1008,
        "firstName": "Nicholas",
        "lastName": "Runolfsdottir",
        "age": 21,
        "gender": "Male",
        "date": "2021-11-07"
    },
    {
        "id": 1021,
        "firstName": "Leroy",
        "lastName": "Blair",
        "age": 21,
        "gender": "Male",
        "date": "2021-11-07"
    },
    {
        "id": 1027,
        "firstName": "Alanna",
        "lastName": "Cantrell",
        "age": 65,
        "gender": "Female",
        "date": "2021-11-07"
    },
    {
        "id": 1016,
        "firstName": "Osbert",
        "lastName": "Valdez",
        "age": 32,
        "gender": "Male",
        "date": "2021-11-02"
    },
    {
        "id": 1024,
        "firstName": "Emmett",
        "lastName": "Marshall",
        "age": 45,
        "gender": "Male",
        "date": "2021-11-02"
    }
];

export const prescriptionData = [
    {
        "brandName": "Tylenol",
        "genericName": "Acetaminophen",
        "dosage": "3 times a day",
        "date": "2021-10-26"
    },
    {
        "brandName": "Advil",
        "genericName": "Ibuprofen",
        "dosage": "3 times a day",
        "date": "2021-10-26"
    },
    {
        "brandName": "Dolo",
        "genericName": "Paracetamol",
        "dosage": "1 time a day",
        "date": "2021-10-26"
    }
];

export const medicineList = [
    {
        "id": 1,
        "brandName": "Dolo",
        "cause": "Incompatible with patients' medical conditions"
    },
    {
        "id": 13,
        "brandName": "Vibramycin",
        "cause": ""
    },
    {
        "id": 17,
        "brandName": "Vibramycin",
        "cause": ""
    },
    {
        "id": 22,
        "brandName": "Betadine",
        "cause": ""
    },
    {
        "id": 25,
        "brandName": "Iodex",
        "cause": ""
    },
    {
        "id": 46,
        "brandName": "Desyrel",
        "cause": "Incompatible with patients' medical conditions"
    }
];