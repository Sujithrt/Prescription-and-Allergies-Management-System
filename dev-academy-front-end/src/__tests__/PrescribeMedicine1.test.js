import React from 'react';
import { shallowToJson } from "enzyme-to-json";
import { shallow } from 'enzyme';
import PrescribeMedicine from '../components/prescribeMedicine/PrescribeMedicine'
import { patient, prescription, medicineList } from '../testConstants'
import axios from "axios"
import MockAdapter from "axios-mock-adapter"

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: () => ({
        pathname: "localhost:3000/patient/1001"
    })
}));

describe("PrescribeMedicine Test", () => {
    it("PrescribeMedicine is rendered as expected", () => {
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        expect(shallowToJson(prescribeMedicine)).toMatchSnapshot();
    });
});

describe("User input Test", () => {
    it("Medicine change with user input", () => {
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        const medicine = {label: "Dolo", value: 1, cause: "Causes MILD Tiredness in patient"}
        prescribeMedicine.find({ 'data-testid': "medicine" }).simulate('change', { target: { 'data-testid': 'medicine', value: medicine } })
        expect(prescribeMedicine.find({ 'data-testid': "medicine" }).prop('value').target.value).toEqual(medicine);
    });
    it("Dose change with user input", () => {
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        prescribeMedicine.find({ 'data-testid': "doseField" }).simulate('change', { target: { 'data-testid': 'dose', value: '1 time a day' } })
        expect(prescribeMedicine.find({ 'data-testid': "doseField" }).prop('value')).toEqual('1 time a day');
    });
});

describe("test add medicine field", () => {
    it('click + button to add new field', () => {
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        prescribeMedicine.find({ 'data-testid': 'addMedicineField' }).simulate('click')
    });
});

describe("test delete medicine field", () => {
    it('click delete button to remove medicine field', () => {
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        prescribeMedicine.find({ 'data-testid': 'deleteMedicineField' }).simulate('click')
    });
});

describe("Prescribe Medicine in new encounter", () => {
    let mock;

    beforeAll(() => {
        mock = new MockAdapter(axios);
    });

    afterEach(() => {
        mock.reset();
    });
    it('All fields filled', () => {
        mock.onPost().reply(201, prescription)
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        const medicine = {label: "Dolo", value: 1, cause: "Causes MILD Tiredness in patient"}
        prescribeMedicine.find({ 'data-testid': "medicine" }).simulate('change', { target: { 'data-testid': 'medicine', value: medicine } })
        prescribeMedicine.find({ 'data-testid': "doseField" }).simulate('change', { target: { 'data-testid': 'dose', value: '1 time a day' } })
        prescribeMedicine.find({ 'data-testid': "prescribeButton" }).simulate('click', { preventDefault: () => { } })
    });
    it('Some fields filled', () => {
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        const medicine = {label: "Dolo", value: 1, cause: "Causes MILD Tiredness in patient"}
        prescribeMedicine.find({ 'data-testid': "medicine" }).simulate('change', { target: { 'data-testid': 'medicine', value: medicine } })
        prescribeMedicine.find({ 'data-testid': "doseField" }).simulate('change', { target: { 'data-testid': 'dose', value: '' } })
        prescribeMedicine.find({ 'data-testid': "prescribeButton" }).simulate('click', { preventDefault: () => { } })
    });
    it('Invalid input', () => {
        mock.onPost().reply(400)
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        const medicine = {label: "Dolo", value: 1, cause: "Causes MILD Tiredness in patient"}
        prescribeMedicine.find({ 'data-testid': "medicine" }).simulate('change', { target: { 'data-testid': 'medicine', value: medicine } })
        prescribeMedicine.find({ 'data-testid': "doseField" }).simulate('change', { target: { 'data-testid': 'dose', value: '1 time a day' } })
        prescribeMedicine.find({ 'data-testid': "prescribeButton" }).simulate('click', { preventDefault: () => { } })
    });
    it('Medicine already prescribed', () => {
        mock.onPost().reply(409)
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        const medicine = {label: "Dolo", value: 1, cause: "Causes MILD Tiredness in patient"}
        prescribeMedicine.find({ 'data-testid': "medicine" }).simulate('change', { target: { 'data-testid': 'medicine', value: medicine } })
        prescribeMedicine.find({ 'data-testid': "doseField" }).simulate('change', { target: { 'data-testid': 'dose', value: '1 time a day' } })
        prescribeMedicine.find({ 'data-testid': "prescribeButton" }).simulate('click', { preventDefault: () => { } })
    });
});

describe("Medicine Select component", () => {
    let mock;

    beforeAll(() => {
        mock = new MockAdapter(axios);
    });

    afterEach(() => {
        mock.reset();
    });
    it('test onInputChange of medicine search', () => {
        mock.onGet().reply(200, medicineList)
        const prescribeMedicine = shallow(<PrescribeMedicine patient={patient} />);
        prescribeMedicine.find({ 'data-testid': 'medicine' }).simulate('inputChange')
    });
});