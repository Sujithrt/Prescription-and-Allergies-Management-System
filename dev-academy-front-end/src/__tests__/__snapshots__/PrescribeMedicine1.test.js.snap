// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrescribeMedicine Test PrescribeMedicine is rendered as expected 1`] = `
<Fragment>
  <span
    className="userUpdateTitle"
  >
    Medicines
  </span>
  <WithStyles(ForwardRef(IconButton))
    data-testid="addMedicineField"
    onClick={[Function]}
  >
    <Memo(ForwardRef(AddIcon)) />
  </WithStyles(ForwardRef(IconButton))>
  <br />
  <form
    onSubmit={[Function]}
  >
    <Container
      fluid={false}
    >
      <Row
        key="0"
        noGutters={false}
      >
        <ForwardRef
          className="search-bar"
          classNamePrefix="select"
          components={
            Object {
              "Option": [Function],
            }
          }
          data-testid="medicine"
          isClearable={true}
          name="medicines"
          onChange={[Function]}
          onInputChange={[Function]}
          options={Array []}
          placeholder="Select medicine..."
          value=""
        />
        <ForwardRef(TextField)
          className="doseInput"
          data-testid="doseField"
          label="Dose"
          name="dose"
          onChange={[Function]}
          value=""
          variant="standard"
        />
        <div
          style={
            Object {
              "marginLeft": "10px",
              "marginTop": "10px",
            }
          }
        >
          <Button
            active={false}
            data-testid="deleteMedicineField"
            disabled={false}
            onClick={[Function]}
            variant="danger"
          >
            Delete
          </Button>
        </div>
      </Row>
      <Button
        active={false}
        data-testid="prescribeButton"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginTop": "10px",
          }
        }
        variant="primary"
      >
        Prescribe
      </Button>
    </Container>
  </form>
</Fragment>
`;
